### ISSUE TITLE
Multiple issues when swapping type variables

### ISSUE THREAD

**Issue by comex** [microsoft/pyright#5373]
**Describe the bug**
- Test 1 sends pyright into a seemingly infinite loop.  It should produce a type error because the type annotation is wrong.
- Test 2 and Test 3 both have code inside a class `TestN[T, U]` that tries to create an instance of the same class but with the type parameters swapped.  These both should succeed as far as I can tell, but produce type errors.
 
**To Reproduce**
Run `pyright` on the below tests.

Result:

**Code or Screenshots**
**Test 1:**
```python
from dataclasses import dataclass
from typing import Generic, TypeVar
T = TypeVar('T')
U = TypeVar('U')

class Test1(Generic[T, U]):
    def __init__(self, t: T, u: U):
        pass

    def test1(self, ts: list[T], us: list[U]) -> None:
        _: Test1[U, T] = Test1(us, ts)
```
**Test 2:**
```python
from dataclasses import dataclass
from typing import Generic, TypeVar
T = TypeVar('T')
U = TypeVar('U')

class Test2(Generic[T, U]):
    def __init__(self):
        pass

    def test2(self) -> None:
        _: Test2[U, T] = Test2[U, T]()
```
**Test 3:**
```python
from dataclasses import dataclass
from typing import Generic, TypeVar
T = TypeVar('T')
U = TypeVar('U')

class Test3(Generic[T, U]):
    def __init__(self, ts: list[T], us: list[U]):
        pass

    def test3(self, ts: list[T], us: list[U]) -> None:
        _: Test3[U, T] = Test3(us, ts)
```
**VS Code extension or command-line**

pyright 1.1.315, command line

**Comment by cicada0007** [comment]
Update `Pyright`: Ensure that you're using the latest version of `Pyright`. You can update it by running the following command:

```
npm install -g pyright

**Comment by erictraut** [comment]
Thanks for the excellent bug report. Pyright 1.1.316 includes a fix for Test2, but it doesn't yet address Test1 or Test3.

**Comment by erictraut** [comment]
This has turned out to be a very difficult bug to fix given the way that pyright's constraint solver is designed. I've made several attempts, but all of them have exploded in complexity and proven to be infeasible. I'll continue to look for solutions. Just wanted to provide an update and explain why this bug has been open for so long.

**Comment by erictraut** [comment]
I think I've found a good fix for this problem. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
